File: Python_programs/lab2_practical_dynamicmemory.py
def output_array(arr):
    print("Array elements:", end=" ")
    for element in arr:
        print(element, end=" ")
    print()

def linear_search(arr, n):
    flag = False
    for i, element in enumerate(arr):
        if element == n:
            print(f"Entered element {element} found at index {i}")
            flag = True

    if not flag:
        print(f"Entered element {n} not found in the array")

def reverse_array(arr):
    reversed_arr = arr[::-1]
    print("Reversed array:", end=" ")
    output_array(reversed_arr)

def update_array(arr):
    updated_arr = [x * 2 if i % 2 == 0 else x + 5 for i, x in enumerate(arr)]
    print("Updated array:", end=" ")
    output_array(updated_arr)

def insert_element(arr):
    index = int(input("Enter index you want to insert element in: "))
    element = int(input("Enter element you want to insert: "))

    arr.insert(index, element)

    print("Array after insertion:")
    output_array(arr)

def delete_element(arr):
    index = int(input("Enter index to be deleted: "))

    if 0 <= index < len(arr):
        del arr[index]
        print("Array after deletion:")
        output_array(arr)
    else:
        print("Invalid index")

def main():
    size = int(input("Enter size of array: "))
    arr = []

    print("Enter elements in the array:")
    for _ in range(size):
        element = int(input())
        arr.append(element)

    output_array(arr)

    while True:
        print("\nMenu:")
        print("1. Linear Search")
        print("2. Reverse Array")
        print("3. Update Array")
        print("4. Insert Element")
        print("5. Delete Element")
        print("6. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            n = int(input("Enter n: "))
            linear_search(arr, n)
        elif choice == 2:
            reverse_array(arr)
        elif choice == 3:
            update_array(arr)
        elif choice == 4:
            insert_element(arr)
        elif choice == 5:
            delete_element(arr)
        elif choice == 6:
            print("Exiting program.")
            break
        else:
            print("Invalid choice. Please enter a valid option.")

if __name__ == "__main__":
    main()

File: C_programs/lab2_practical_dynamicmemory.c
#include <stdio.h>
#include<stdlib.h>

void output_array(int a[], int size) {
    printf("Array elements: ");
    for (int i = 0; i < size; i++) {
        printf("%d ", a[i]);
    }
    printf("\n");
}

void linear_search(int a[], int size) {
    int n;
    printf("\nEnter n: ");
    scanf("%d", &n);

    int flag = 0; 
    for (int i = 0; i < size; i++) {
        if (a[i] == n) {
            printf("Entered element %d found at index %d\n", a[i], i);
            flag = 1; 
        }
    }

    if (flag == 0) {
        printf("Entered element %d not found in the array\n", n);
    }
}

void reverse_array(int a[], int size) 
{

    int i = 0, j = size - 1;

    while (j > i) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
        i++;
        j--;
    }

    printf("Reversed array: ");
    output_array(a, size);
}

void update_array(int a[], int size) {
    printf("Updated array: ");
    for (int i = 0; i < size; i++) 
    {
        if (i % 2 == 0)
            a[i] *= 2;
        else
            a[i] += 5;
    }

    output_array(a, size);
}

void insert_element(int **a, int *size ) {
    int index, element;
    printf("Enter index you want to insert element in: ");
    scanf("%d", &index);

    printf("Enter element you want to insert: ");
    scanf("%d", &element);

    (*size)++;
    *a = realloc(*a,(*size)*sizeof(int));

    for (int i = *size - 1; i > index; i--) {
        (*a)[i] = (*a)[i - 1];
    }
    (*a)[index] = element;

    printf("Array after insertion:\n");
    output_array(*a, *size);
}

void delete_element(int **a, int *size) {
    int index;
    printf("Enter index to be deleted: ");
    scanf("%d", &index);

    if (index < 0 || index >= *size) 
    {
        printf("Invalid index\n");
        return;
    }

    for (int i = index; i < *size - 1; i++) 
    {
        (*a)[i] = (*a)[i + 1];
    }

    (*size)--;

    printf("Array after deletion:\n");
    output_array(*a, *size);
}

int main(void) 
{
    int size;
    printf("Enter size of array : ");
    scanf("%d",&size);

    int* a = (int*)malloc(size*sizeof(int));
    
    printf("Enter elements in the array: ");
    for (int i = 0; i < size; i++) 
    {
        scanf("%d", &a[i]);
    }

    output_array(a, size);

    int choice;

    do {
        printf("\n Menu:\n");
        printf("1. Linear Search\n");
        printf("2. Reverse Array\n");
        printf("3. Update Array\n");
        printf("4. Insert Element\n");
        printf("5. Delete Element\n");
        printf("6. Exit\n");

        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                linear_search(a, size);
                break;
            case 2:
                reverse_array(a, size);
                break;
            case 3:
                update_array(a, size);
                break;
            case 4:
                insert_element(&a, &size);
                break;
            case 5:
                delete_element(&a,&size);
                break;
            case 6:
                printf("Exiting program.\n");
                break;
            default:
                printf("Invalid choice. Please enter a valid option.\n");
                break;
        }

    } 
    while (choice != 6);

    free(a);
}